dask/bytes/core.py

```diff
@@ -281,64 +223,66 @@ class OpenFile(object):
 
     Parameters
     ----------
-    myopen: function
-        Opens the backend file. Should accept path and mode, as the builtin open
-    path: str
+    fs : FileSystem
+        The file system to use for opening the file
+    path : str
         Location to open
-    compression: str or None
-        Compression to apply
-    mode: str like 'rb'
+    mode : str like 'rb', optional
         Mode of the opened file
-    text: bool
+    compression : str or None, optional
+        Compression to apply
+    text : bool, optional
         Whether to wrap the file to be text-like
-    encoding: if using text
-    errors: if using text
+    encoding : str or None, optional
+        The encoding to use if opened in text mode.
+    errors : str or None, optional
+        How to handle encoding errors if opened in text mode.
     """
-    def __init__(self, myopen, path, compression, mode, text, encoding,
+    def __init__(self, fs, path, mode='rb', compression=None, encoding=None,
                  errors=None):
-        self.myopen = myopen
+        self.fs = fs
         self.path = path
-        self.compression = compression
         self.mode = mode
-        self.text = text
+        self.compression = get_compression(path, compression)
         self.encoding = encoding
-        self.closers = None
-        self.fobjects = None
         self.errors = errors
-        self.f = None
+        self.fobjects = []
+
+    def __reduce__(self):
+        return (OpenFile, (self.fs, self.path, self.mode, self.compression,
+                           self.encoding, self.errors))
 
     def __enter__(self):
         mode = self.mode.replace('t', '').replace('b', '') + 'b'
-        f = f2 = self.myopen(self.path, mode=mode)
-        CompressFile = merge(seekable_files, compress_files)[self.compression]
-        if PY2:
-            f2 = SeekableFile(f)
-        f3 = CompressFile(f2, mode=mode)
-        if self.text:
-            f4 = io.TextIOWrapper(f3, encoding=self.encoding,
-                                  errors=self.errors)
-        else:
-            f4 = f3
 
-        self.closers = [f4.close, f3.close, f2.close, f.close]
-        self.fobjects = [f4, f3, f2, f]
-        self.f = f4
-        f4.close = self.close
-        return f4
+        f = SeekableFile(self.fs.open(self.path, mode=mode))
```

Comment:
SeekableFile used to be under a PY2 branch.  This is no longer necessary?

Reply:
Can we guarantee that the file-like produced with `self.fs.open(self.path, mode=mode)` will have its `close()` called in every case?

Reply:
`SeekableFile` is either the identity, or forwards the call to close. Close of the output file-like is always called, so this is fine.